### PR TITLE
Cleanup/update model inversion notebook

### LINKED ISSUE #201: Update example notebooks to handle batch size parameter
based on https://github.com/infer-actively/pymdp/issues/173 and github.com/infer-actively/pymdp/pull/186

### PR DESCRIPTION
**Pull request by conorheins**
To address #242 (and thus close #201), #287, #295

Updates the `examples/model_fitting` notebook with the updates to how the `batch_size` argument is passed to the `Agent` constructor, among other minor API updates (like the structure of the return of `self.reset()` method of `pymdp.envs.Env`).

### Summary of changes
- add an optional `modelfit` dependency group to the `pyproject.toml`, which adds the `pybefit` package as a dependency. For the time being (until these changes are merged to `v1.0.0_alpha` and then to `master`), this is pointing to a branch of `pybefit` called `fix/pymdp-compat` on the `conorheins` fork.
- remove deprecated `model_inversion.ipynb` notebook which assumed deprecated APIs of both `pymdp` and `pybefit`, and replaced it with a fresh new notebook called `fitting_with_pybefit.ipynb`, which fully depends on `pybefit`'s custom likelihood functions for `pymdp`. See discussion in thread of #281 and #287 for context / motivation for this decision. This notebook is currently running without errors or NaNs, but the parameter recovery plots are not good (no correlation between ground truth parameter values and the inferred ones), but this is likely an issue with the model identifiability rather than the functioninig of `pybefit`. It uses the most up-to-date version of the TMaze task. The `pybefit` package itself has an `examples/pymdp/` folder with a functioning model inversion example (see [here](https://github.com/conorheins/pybefit/blob/fix/pymdp-compat/examples/pymdp/simulating_and_fitting_agency_task.ipynb), based on a real [agency experiment](https://osf.io/preprints/osf/ckfrt) from computational psychology (by @riddhipits and @dimarkov).
- adds unit tests for jittability of `__init__` of `Agent` (needed for compatibility with `pybefit`) and for autodifferentiability of: building an agent and running a single timestep of active inference , addressing #295. Also refactored the normalization checking function in `utils.py` to use an equinox `error_if` function which allows graceful handling of runtime errors in JIT context.

### CODE REVIEW

File: test/test_agent_jax.py [hunk 1]
```diff
@@ -490,6 +494,205 @@ def test_agent_with_A_learning_requires_pA(self):
...
+            return agent
+        
+        # write a unit test that the Agent constructor is jittable
+        jitted_constructor = jit(construct_agent)
+        agent_instance = jitted_constructor(
+            A_batched, B_batched, pA_batched, pB_batched, D_batched
+        )
+        self.assertIsInstance(agent_instance, Agent)
+
+    def test_valid_gradients_one_step_ahead(self):
```
Comment (Arun-Niranjan) [inline test/test_agent_jax.py:557]: At the moment I believe this test will fail at the first exception, which means we would lose coverage on the subsequent ones. It probably makes sense to break this test out into parameterized sub tests - one for each inference algorithm, with the added benefit of being able to run each version of the test in parallel.
https://docs.pytest.org/en/stable/how-to/parametrize.html
But that's something I will take a look at separately

**Review by Arun-Niranjan — COMMENTED** [review]
I can't check the notebook from my phone, but the rest of the changes + the tests look good to me.

File: test/test_agent_jax.py (hunk 1 above)
Comment (conorheins) [reply test/test_agent_jax.py:557]: Good point - i was being lazy and just enumerated the cases sequentially, but you're right we should absolutely parameteriz pytests like this (there are others that deserve similar treatment throughout the test suite). I've done this before and it's pretty straightforward. thanks for offering to tackle it
